attestation-container/client-example/client-example.go

```diff
@@ -0,0 +1,380 @@
+	r.Measurement = hex.EncodeToString(report[144:192])
+	r.HostData = hex.EncodeToString(report[192:224])
+	r.IDKeyDigest = hex.EncodeToString(report[224:272])
+	r.AuthorKeyDigest = hex.EncodeToString(report[272:320])
+	r.ReportID = hex.EncodeToString(report[320:352])
+	r.ReportIDMA = hex.EncodeToString(report[352:384])
+	r.ReportedTCB = binary.LittleEndian.Uint64(report[384:392])
+	r.Reserved2 = hex.EncodeToString(report[392:416])
+	r.ChipID = hex.EncodeToString(report[416:480])
+	r.CommittedSvn = binary.LittleEndian.Uint64(report[480:488])
+	r.CommittedVersion = binary.LittleEndian.Uint64(report[488:496])
+	r.LaunchSvn = binary.LittleEndian.Uint64(report[496:504])
+	r.Reserved3 = hex.EncodeToString(report[504:672])
+	r.Signature = hex.EncodeToString(report[672:1184])
+
+	return nil
+}
+
+func reverseBytes(bytes []byte) []byte {
+	// Deep copy to avoid side effects on the parameter
+	out := append(make([]byte, 0, len(bytes)), bytes...)
+	for i := 0; i < len(out)/2; i++ {
+		j := len(out) - i - 1
+		out[i], out[j] = out[j], out[i]
+	}
+	return out
+}
+
+// Deserialize ECDSA signature with P384 and SHA384
+func deserializeSignature(hexSignature string) (ECDSASignatureP384SHA384, error) {
+	sigBytes, err := hex.DecodeString(hexSignature)
+	if err != nil {
+		log.Fatalf("Failed to decode signature: %s", err.Error())
+	}
+	if len(sigBytes) != SIGNATURE_SIZE {
+		log.Fatalf("Signature length should be %d bytes, but it's %d", SIGNATURE_SIZE, len(sigBytes))
+	}
+	rbytesLittleEndian := sigBytes[0:72]
+	sbytesLittleEndian := sigBytes[72 : 72+72]
+	// Big endian is requred by `SetBytes()`.
+	rbytesBigEndian := reverseBytes(rbytesLittleEndian)
+	sbytesBigEndian := reverseBytes(sbytesLittleEndian)
+	return ECDSASignatureP384SHA384{
+		R: new(big.Int).SetBytes(rbytesBigEndian),
+		S: new(big.Int).SetBytes(sbytesBigEndian),
+	}, nil
+}
+
+type UVMEndorsementsPayload struct {
+	GuestSVN          string `json:"x-ms-sevsnpvm-guestsvn"`
+	LaunchMeasurement string `json:"x-ms-sevsnpvm-launchmeasurement"`
+}
+
+func fetchAttestation() *pb.FetchAttestationReply {
+	// Set up a connection to the server.
+	dialer := func(addr string, t time.Duration) (net.Conn, error) {
+		return net.Dial("unix", addr)
+	}
+	conn, err := grpc.Dial(*addr, grpc.WithInsecure(), grpc.WithDialer(dialer))
+	if err != nil {
+		log.Fatalf("Did not connect: %v", err)
+	}
+	defer conn.Close()
+	c := pb.NewAttestationContainerClient(conn)
+
+	// Contact the server and print out its response.
+	ctx, cancel := context.WithTimeout(context.Background(), TIMEOUT_IN_SEC*time.Second)
+	defer cancel()
+	// public key bytes in UTF-8 (https://go.dev/blog/strings)
+	publicKey := []byte("public-key-contents")
+	r, err := c.FetchAttestation(ctx, &pb.FetchAttestationRequest{ReportData: publicKey})
+	if err != nil {
+		log.Fatalf("Could not get attestation: %v", err)
+	}
+	return r
+}
+
+func verifyAttestationEndorsements(endorsementCertificates []byte) *x509.Certificate {
```

Comment:
Endorsement is a bit overused. In the MAA case that means the UVM reference info. I think you are verifying the platform certs here?

Reply:
I followed CCF's convention.

https://github.com/microsoft/CCF/blob/main/attestation-container/protobuf/attestation-container.proto#L20

attestation endorsements: endorses attestation report
UVM endorsements: endorses UVM

>  I think you are verifying the platform certs here

Yes, I am. So it can be `verifyPlatformCertificats` or something like that, but in that case, it might bit confusing considering that server and client are using different names for a same thing.
(of course, we can rename the interface carefully though in future)

Reply:
I think you make a good point. Leave it so it matches CCF. Maybe some comment block to make it clear locally which set of endorsements are which. AttestationEndorsement is true when you think about it but for someone coming to the code without a deep understanding of what is going on it might be a bit of a leap. We can help them by explaining that the attestation endorsement is the matching certs from AMD, possibly supplied in a round about way etc etc...

Reply:
Happy to switch naming on the CCF side to better match conventions used elsewhere.